DOC: Change recommendation away from pinning numpy+3

stefanv opened the pull request:
SciPy has a policy of pinning to `numpy<=current_ver + 3`.  While this may be fine for SciPy (or, put otherwise, for exactly *one* package in the ecosystem), I think a proliferation of this habit will certainly cause problems.

Imagine many projects started pinning to numpy+k. If any one of these projects is then unable to make a release in time, it can easily make it impossible to depend on a combination of them.

Pip isn't very strict about enforcing upper limits at the moment, but it complains if you try to build numpy in an environment with a not-very-old (<1.8) scipy installed—which happens commonly, since scipy is a dependency of the numpy docs.

A possible recommendation could be to warn if the package thinks that NumPy may be too old.  But with NumPy being so conservative in changing its API, forward-pinning causes more problems than it solves.

Also, if we urge users to pin to `<next_major_version`, we always have that mechanism of control (from the NumPy developer side) in case we want to change something that is likely to break multiple packages.  But some will even debate whether we want this pin in place (it will force all dependencies to "sign off" on a major NumPy release).

Also see https://iscinumpy.dev/post/bound-version-constraints/

stefanv committed:
Change recommendation away from pinning numpy+3

tylerjereddy commented:
In case it helps, I *think* downstream we may be "ok" with say `++max_numpy_version` if a new NumPy comes out and it happens to not break anything while we're in the process of patching a release series anyway. That was the sense I got from the SciPy meeting this morning.

I know you're aiming for broader advice in any case, and the debates seem a bit complex, but just wanted  to mention that I believe SciPy will make an effort to bump a bit more when we can, though obviously doing this manually and probably when we just happen to be patching for some other reason may not quite get what you're looking for, but perhaps better than complete rigidity.

stefanv commented:
> I know you're aiming for broader advice in any case, and the debates seem a bit complex, but just wanted to mention that I believe SciPy will make an effort to bump a bit more when we can, though obviously doing this manually and probably when we just happen to be patching for some other reason may not quite get what you're looking for, but perhaps better than complete rigidity.

Thanks, Tyler, I appreciate that SciPy will try to be as lenient as deemed comfortable. Do you have a sense of how often new releases of NumPy break SciPy?

rgommers reviewed (changes requested):
@stefanv the docs for what you are talking about are at https://numpy.org/devdocs/user/depending_on_numpy.html#adding-a-dependency-on-numpy

> Imagine many projects started pinning to numpy+k. If any one of these projects is then unable to make a release in time, it can easily make it impossible to depend on a combination of them.

This is why the docs emphasize to only do this for well-maintained packages with a regular release schedule. Those are SciPy, scikit-learn, scikit-image, AstroPy etc. I don't think there are more than O(10-100) of such packages that are really widely used. If you think that the docs aren't explicit enough, can you suggest changes?

> Thanks, Tyler, I appreciate that SciPy will try to be as lenient as deemed comfortable. Do you have a sense of how often new releases of NumPy break SciPy?

Pretty much guaranteed that something breaks in every single minor release. The question is more if it's for a few functions or if there is widespread breakage (e.g., no more indexing with floats breaks things all over the place). There are also changes something that break importing a submodule or all of SciPy. Due to that, the value of the pin for SciPy isn't really in question imho. And the same is true for scikits.

> While this may be fine for SciPy (or, put otherwise, for exactly _one_ package in the ecosystem), I think a proliferation of this habit will certainly cause problems.

No, I'm not willing to remove that recommendation at this point. The recommendations for both runtime and build-time pins were designed with a lot of care. Some PyPA members repeating the discussion over and over (which is mostly aimed at Poetry auto-adding pins, not at us) is getting annoying.

rgommers commented:
To add: we really need solid evidence of real-world problems. I've had extensive discussions on this, and a hypothtical "may cause problems" isn't enough and won't really add much to the discussions and bug reports that preceded writing these docs.

rgommers commented:
> Ralf, did you maybe confuse this PR for an issue?

I did in my first sentence (although that link is still helpful, so I left it).

For the rest: no I didn't. I suggest to look at concrete issues if you have them. I do not agree with this PR as it stands. You're trying to change something that I designed with quite a bit of care without pointing at concrete issues.

stefanv commented:
Cf. https://github.com/numpy/numpy/pull/18505

The PR did not have much discussion. It may be worth pinging the mailing list to get a sense of what others think.

I appreciate that you took care in writing the document, and for the most part it is excellent. But I think that this specific recommendation is flawed in principle, as illustrated by the article linked to and by reasoning about  implications.

The reason you do not see much concrete evidence of problems yet is because few packages other than SciPy is following the advice. But we can avert those concrete problems by te-considering the document now.

rgommers commented:
> But I think that this specific recommendation is flawed in principle, as illustrated by the article linked to and by reasoning about its implications.

The blog writes:

_I [have seen](https://numpy.org/devdocs/user/depending_on_numpy.html#runtime-dependency-version-ranges) a +3 rule for capping proposed by NumPy. If your code works without warnings on the current NumPy version 1.x, you should cap to 1.(x+3). This is better; you can test development versions, you can avoid version bound clashes for a while, but still unnecessary for most code - you are very unlikely to be using something in NumPy that is about to be removed. It’s better to simply acknowledge that if you use a library years after it releases, you might need to manually control some of the versions. If you vanish of the face of the internet, then an uncapped library will be more compatible than a wrongly capped one, and if you are still around, you’ll keep it up to date. If you heavily use NumPy, though, this could be an acceptable cap. It’s also much less of a problem if you have a large library with multiple maintainers and regular releases._

_Also, my focus here has been on smaller libraries; if you have a dozen developers working on the project, I’m assuming you already have a reasonable (anti) capping policy (though, to be fair, TensorFlow made this mistake for a while). A large library that has regular updates would probably be quite reasonable on this +3 rule._

So it's already mildly positive/okay, **while not acknowledging the main reason for the cap** - which is that if breakage is likely, as it is for large projects like SciPy and the main scikits, an install of a version is going to break after 1.5-2 years **and our user base who are primarily not software developers do not know how to deal with that, and shouldn't have to**. 

> But we can avert those concrete problems by re-considering the document now.

You have to be a lot more concrete here Stefan. This policy is designed to avoid a class of problems, and - if executed well - should not introduce serious new problems.

bashtage commented:
I wish statsmodels was better at setting both lower and upper bounds.  We have fewer problems with NumPy, but we often see issues where someone picks up an old statsmodels but wants to have the latest pandas. This combination prevents the standard API entrypoint from being imported.  We test against the nightlies of SciPy, NumPy, and pandas and aggressively squash and try to backport Future and Deprecation Warning.  

I think the recommendation to pin `numpy>={min_ver},<{latest_release+3}` seems fairly prudent.

stefanv commented:
> You have to be a lot more concrete here Stefan. This policy is designed to avoid a class of problems, and - if executed well - should not introduce serious new problems.

I think the concrete example was what I gave in the PR description, so not sure what more you want me to add.  

The primary problem lies with "if executed well": the recommendation should only be followed by packages who are guaranteed to release frequently (and the text says so, but small packages like to emulate bigger ones, so that warning won't be heeded). Smaller packages may genuinely aim to release frequently and still fail.  SciPy was recently held up for months over some minor issue, and perhaps it is crazy to imagine it being held up for several NumPy releases.  But SciPy is arguably one of the best resourced libraries in the ecosystem, and it is not that well resourced.

To be clear: **I am not trying to convince the SciPy developers** to change what they're doing; all I am proposing is to remove advice from the NumPy docs that feels unnecessary and may cause problems.  Reading your quote from the article, it mainly emphasizes that the advice is not suitable for or well applicable to smaller (most) libraries.

>  if breakage is likely, as it is for large projects like SciPy and the main scikits, an install of a version is going to break after 1.5-2 years

This could be addressed by a warning to the user: note, you are using an older version of the library with a newer version of NumPy; if you run into issues, please upgrade. But instead we are now *forcing* users to upgrade when that may very well not be necessary for many libraries.

Also, if things break, at least people tend to upgrade to try and fix things. The ugly spectre to avoid is the so-called Hinsen principle, whereby output changes silently without the API changing—but we are quite careful about not doing that, for the most part.

pradyunsg reviewed (commented):
This is very much a drive by comment. Please don't @ me separately. :)

I just want to call out a nuance that isn't clear from the PR title alone.

asmeurer commented on doc/source/user/depending_on_numpy.rst:
```diff
@@ -114,25 +114,18 @@ for dropping support for old Python and NumPy versions: :ref:`NEP29`. We
-exceeded which _may_ lead to your package being hard to install in combination
-with other packages relying on the latest NumPy.
+``setuptools`` to build). The upper bound should be set to the next
+major release of NumPy, e.g. `numpy>=1.18,numpy<2`.
```
I agree with this. It should be possible to separate the documenting of how versions are numbered with the recommending (if any) of how to pin upstream.

stefanv replied:
NumPy 2.0 is the only release in which we are even contemplating making major breaking changes.  Otherwise, NumPy remains for the greater part backwards compatible, not counting ABI, bug-fixes, and small tweaks.  ABI breakage does not affect most Python packages.

That said, I don't consider this an important part of the PR, so I can remove that part, or adjust it if you have some guidance.

rgommers replied:
> NumPy 2.0 is the only release in which we are even contemplating making major breaking changes

This is not actually true. Two examples that immediately come to mind which deprecated widely used functionality (which then breaks older releases of pretty much every major package, and not by accident):

- Indexing with floats errors: https://numpy.org/devdocs/release/1.12.0-notes.html#deprecationwarning-to-error
- (removal coming soon I think) deprecated `np.int`, `np.float` et al: https://numpy.org/devdocs/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated

I think there are more such large changes, and we're not always opposed to making them - it's a cost/benefit thing. Value-based casting is one we're looking at now for example.

stefanv committed:
Change language to better explain upper bound pinning

This now makes it explicit when upper bounds should be set, and returns previous
guidance on `numpy_version + 3`.

stefanv commented:
I've updated the PR language to try and come to a compromise; making it not the default recommendation, but stating the circumstances under which it can be considered.

rgommers commented:
> It still presumes that libraries are well-maintained having regular releases.

The text literally says that already, so your comment is off-base.

> I've updated the PR language to try and come to a compromise; making it not the default recommendation, but stating the circumstances under which it can be considered.

Thanks @stefanv, this looks a lot more reasonable. If we're going to tweak this, let's maybe also make it more concrete. There's multiple cases actually:
- No upper bounds: if you use a small API surface of NumPy's Python API, don't do regular releases, there's a chance releases will be discontinued, etc.
- `<2.0`: at least needed if you use the NumPy C API (possibly via use of NumPy in Cython code), because 2.0 is quite likely to contain an ABI break. It's also more likely to contain large API breaks. 
- `<1.N+3`: for packages that use a large API surface and have regular release cycles: the likes of SciPy, Pandas, Matplotlib, scikit-learn, scikit-image, AstroPy, etc. (`statsmodels` is a bit of a special case, it'd be nice but it doesn't have a regular enough release cycle, so I'd recommend `N+4/5`).

stefanv committed:
Incorporate C-API and large API surface usage suggestions by Ralf

stefanv commented:
Updated to include the language on C ABI breakage and large API surface usage.

tacaswell commented on doc/source/user/depending_on_numpy.rst:
```diff
@@ -114,26 +114,23 @@ for dropping support for old Python and NumPy versions: :ref:`NEP29`. We
+
+That said, if you are (a) a project that is guaranteed to release
+frequently, (b) use a large part of NumPy's API surface, and (c) is
+worried that changes in NumPy may break your code, you may set an
```
pick one of 'can' and 'may' between this sentence and the major version pinning below?

tacaswell commented:
I went back through old Matplotlib versions and tried to sort out where changes to numpy actually broke Matplotilb:


|mpl version | numpy at release | test warnings | test failures |
| --- | --- | --- | ---- |
|3.2.0 | 1.18 | 1.19 | - |
|3.1.0| 1.16 | 1.17 | - |
|3.0.0| 1.15 | 1.15 | 1.18 | 
|2.2.0 | 1.14 | * | 1.18 | 
|2.1.0 | 1.13 | * | 1.14/1.18 |

(For the 2.x series I had issues finding a version of pytest + plugins that both ran and did not give lots of warnings about pytest / internal warnings so not sure when numpy started warning)

From mpl 3.1 up I think we work (but with warnings) up to numpy 1.22 and 1.18 is the version that has what would be a user-facing failure for older versions (due to `np.expand_dims` getting stricter about inputs).   With mpl 2.1 we had a test failure due to a repr length on a float changing with 1.14 which I want to sweep under the carpet.    In the recent (last 5 years), the proposed pin only would have been "right' once and too restrictive in all other cases.  This may be mostly due to Matplotlib's direct use of numpy actually being a relatively small footprint on the library.

I think a core of this issue is that the versions in `setup.py` / `pyproject.toml` are immutable part of the released version of the library.   If we could change the dependency pinning on pypi without doing a new release, I would be 💯 on the recommendation of reasonably tight upper-bound pinning on that information.


I think the updated wording is a good compromise.

stefanv committed:
Consistent use of may/can as per @tacaswell's feedback

rgommers commented:
Thanks @tacaswell, that's very useful data that must have taken a fair bit of work to produce.

> In the recent (last 5 years), the proposed pin only would have been "right' once and too restrictive in all other cases.

One minor comment on your phrasing here: rather than "right once" I'd say that all releases <=3.0.0 now have failures, so a pin would have been right for all of those. Newer releases show warnings, and those will likely turn into failures in the future - the question is just when. The one reason that they wouldn't turn into failures when the relevant deprecation(s) expire(s) is if the range of supported Python versions between those Matplotlib releases and the NumPy version where the deprecated code is removed have no overlap at all (a ~4 year window).

rgommers commented on doc/source/user/depending_on_numpy.rst:
```diff
@@ -114,26 +114,23 @@ for dropping support for old Python and NumPy versions: :ref:`NEP29`. We
+That said, if you are (a) a project that is guaranteed to release
+frequently, (b) use a large part of NumPy's API surface, and (c) is
+worried that changes in NumPy may break your code, you can set an
+upper bound of `<MAJOR.MINOR + N` with N no less than 3, and
```
double backticks for typesetting as code I think

rgommers commented on doc/source/user/depending_on_numpy.rst:
```diff
@@ -114,26 +114,23 @@ for dropping support for old Python and NumPy versions: :ref:`NEP29`. We
-What to do here depends on your release frequency. Given that NumPy releases
-come in a 6-monthly cadence and that features that get deprecated in NumPy
-should stay around for another two releases, a good upper bound is
-``<1.(xx+3).0`` - where ``xx`` is the minor version of the latest
```
Note: this explanation for why `N+3` should always be safe is actually useful, and is just deleted. We should put it back in a follow-up (see my main review comment).

stefanv replied:
I will add a footnote with that information for now.

rgommers replied:
Thanks, that looks good.

rgommers commented on doc/source/user/depending_on_numpy.rst:
```diff
@@ -114,26 +114,23 @@ for dropping support for old Python and NumPy versions: :ref:`NEP29`. We
+C API (directly or via Cython), you can also pin the current major
+version to prevent ABI breakage. Note that setting an upper bound on
+NumPy may `affect the ability of your library to be installed
+alongside other, older packages
```
"older" should be "newer" I think?

rgommers reviewed (approved):
This looks reasonable and I'm fine with merging as is. I would like to add a more detailed explanation in a follow-up; now that we use `sphinx-panels` I think it is good to do this in a foldout so it doesn't confuse readers or detract from the main message. I don't have time to work on that right now, but can do it in the future.

> This could be addressed by a warning to the user: note, you are using an older version of the library with a newer version of NumPy; if you run into issues, please upgrade. [...]

This kind of demonstrates the problem. There are many reasons why a user wants to use an older version of some package: matching the version in a production pipeline or a paper, for debugging purposes, etc. `pip install pkgname` always works and is the default people use, if a user explicitly types `pip install pkgname==x.y.z` then giving the advice "sorry may not work, please upgrade to latest" is kind of counterproductive. What the user then actually wants is to _downgrade numpy_, to some version that does work. Which they then have to figure out by trial-and-error or by comparing release dates on PyPI.

stefanv commented:
> This looks reasonable and I'm fine with merging as is. I would like to add a more detailed explanation in a follow-up; now that we use `sphinx-panels` I think it is good to do this in a foldout so it doesn't confuse readers or detract from the main message. I don't have time to work on that right now, but can do it in the future.

That sounds good, thank you, Ralf.

> > This could be addressed by a warning to the user: note, you are using an older version of the library with a newer version of NumPy; if you run into issues, please upgrade. [...]
> 
> This kind of demonstrates the problem. There are many reasons why a user wants to use an older version of some package: matching the version in a production pipeline or a paper, for debugging purposes, etc. `pip install pkgname` always works and is the default people use, if a user explicitly types `pip install pkgname==x.y.z` then giving the advice "sorry may not work, please upgrade to latest" is kind of counterproductive. What the user then actually wants is to _downgrade numpy_, to some version that does work. Which they then have to figure out by trial-and-error or by comparing release dates on PyPI.

That's a good point: the advice should be more specific than "update". Since we know the compatible version numbers, we could advise explicitly which NumPy versions to install to make it work.  I.e., you can either upgrade this library, or downgrade NumPy to `>= 1.x, <= 1.y`.

stefanv committed:
Add footnote about release cadence; fix formatting

stefanv commented:
@rgommers Please check the footnote; I tried to capture the timing necessitated by the `n+3` pinning.

rgommers reviewed (approved):
In it goes, thanks @stefanv & all
